CSUB-708: `creditcoin-cli` integration tests for `register-address` and `collect-coins`

zacharyfrederick opened the pull request:
I used this chance to experiment with some syntax changes for the `globalSetup.ts` file in the `cc-cli` repo. If we like them I will refactor the `integration-tests` setup file to use a similar syntax. 

The goal was to replace these calls. 

```typescript
if ((global as any).CREDITCOIN_CREATE_SIGNER === undefined) {
      (global as any).CREDITCOIN_CREATE_SIGNER = createSigner; // eslint-disable-line
}

if ((global as any).CREDITCOIN_API_URL === undefined) {
    const wsPort = process.env.CREDITCOIN_WS_PORT || '9944';
    (global as any).CREDITCOIN_API_URL = `ws://127.0.0.1:${wsPort}`;
}
```

To facilitate that I added a `globalDefaults` map <string, any> that allows you to define the environment variables needed for a testing environment next to their defaults in a single location. This is meant to be self-documenting. 

```typescript
const globalDefaults = new Map<string, any>([
    ["CREDITCOIN_EXECUTE_SETUP_AUTHORITY", true],
    ["CREDITCOIN_CREATE_SIGNER", createSigner],
    ["CREDITCOIN_ETHEREUM_NODE_URL", "http://127.0.0.1:8545"],
    [
        "CREDITCOIN_API_URL",
        `ws://127.0.0.1:${process.env.CREDITCOIN_WS_PORT ? process.env.CREDITCOIN_WS_PORT : "9944"
        }`,
    ],
]);
```

Then within the setup call, they all get initialized using a `setArgIfUndef` function

```typescript
  globalDefaults.forEach((value: any, key: string) => {
      setArgIfUndef(key, value);
  });
```

You can call `setArgIfUndef` or `setArg` manually if you want. Like here where we set arguments manually after deploying the Gluwa Gate token.

```typescript
await deployCtcToken((global as any).CREDITCOIN_CTC_CONTRACT_ADDRESS);
setArg(
    "CREDITCOIN_CTC_CONTRACT_ADDRESS",
    process.env.CREDITCOIN_CTC_CONTRACT_ADDRESS,
);
setArgIfUndef(
    "CREDITCOIN_CTC_BURN_TX_HASH",
    process.env.CREDITCOIN_CTC_BURN_TX_HASH,
);
```

From within your tests, you can use the `arg` helper function to access the values from within tests 

```typescript
const result = execa.commandSync(
    `npx creditcoin-cli register-address -u ${arg("CREDITCOIN_API_URL")}`,
    {
        env: {
            BLOCKCHAIN: "Ethereum",
            PRIVATE_KEY: ethWallet.privateKey,
            CC_SECRET: caller.secret,
        },
    },
);
```

Also, everyone should link whatever flags they add to the CLI to an environment variable. 

```typescript
  const privateKeyOpt = new Option(
    "-p, --private-key <key> The private key for the address that you want to register.",
  ).env("PRIVATE_KEY");
```

@pLabarta Your functions to fund accounts and sign and watch transactions were super super helpful. I had to make a tiny change to the sign and watch function after I added the authorities though. For the integration tests, the global setup function sets authorities and funds those accounts using normal sudo calls from Polkadot. If you try to sudo fund an account close to the time the authorities are being set up, you can end up trying to insert lower-priority transactions into the pool and the calls will fail.  I only had to add a nonce value of a negative one to the `signAndSend` call and it all worked properly.

Commit:
added global setup functionality

Commit:
added collect-coins and register-address integration tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

atodorov commented on scripts/cc-cli/src/commands/collectCoins.ts:
```diff
@@ -20,7 +20,17 @@ export function makeCollectCoinsCmd() {
     .addOption(externalAddressOpt)
     .addOption(burnTxHashOpt)
-    .action(collectCoinsAction);
+    .action(collectCoinsActionSync);
```
Is this an intended change? Looks like it was meant only for debugging purposes ?

zacharyfrederick replied:
With the previous async function the process would hang after even after completing the collect coins validation. For the CLI only that isn't a major issue because you can just `ctrl-c` out of it but the `execa` command runner blocks until the child process exits. Rewriting it in this synchronous fashion made it a little bit easier to ensure we got correct and easily testable behavior

atodorov commented on scripts/cc-cli/src/commands/registerAddress.ts:
```diff
@@ -52,7 +52,7 @@ async function registerAddressAction(options: OptionValues) {
 function handleSuccess(addressRegistered: AddressRegistered) {
   console.log(
     chalk.green(
-      `Address Registered Successfully!: ${addressRegistered.itemId}`,
```
Why remove ?

zacharyfrederick replied:
I removed it to assert that the final message of stdout was `Address Registered Successfully!`. In hindsight, I should have kept the output the same and checked using a `substring` method. I will revert that change so the output is more useful to our users.

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,116 @@
+import { AUTHORITY_SURI } from "creditcoin-js/lib/examples/setup-authority";
+import { testData, tryRegisterAddress } from "creditcoin-js/lib/testUtils";
+
+describe("collect-coins", () => {
```
integration-tests/src/utils.ts defines a `describeIf` function which will skip the entire section if you can't execute the sudo transaction. Will save you some ifconditions down below.

zacharyfrederick replied:
Any opposition to me moving `testIf` and `describeIf` into `creditcoin-js`?

atodorov replied:
No, however that's probably best to be a separate PR as it will introduce other unrelated changes after itself.

atodorov reviewed (changes requested):
Questions and minor changes requested. Overall looks good.

Commit:
added a step to start local ethereum node to the integration-test-cli workflow

Commit:
linted and formatted

Commit:
updated formatting

Commit:
formatted

atodorov commented on scripts/cc-cli/src/globalSetup.ts:
```diff
@@ -0,0 +1,149 @@
+  if (process.env.INTEGRATION_TEST === undefined) {
+    console.log("Skipping token deployment and authority setup");
+    return;
+  }
```
I don't understand the role of the `INTEGRATION_TEST` env variable.  

jest is already executing the integration test suite and we don't need yet another way to tell it that it is already doing so. What's the purpose here  ? 

Perhaps the deployment of CTC contract can be moved into `beforeAll` as part of collectCoins.test.ts.  it doesn't seem that it is needed elsewhere.

zacharyfrederick replied:
It is unnecessary in retrospect. I moved the token deployment to the before-all method of the collect coins test. Originally I was copying the testing setup from the integration tests which is why it was in the global setup.

atodorov commented on scripts/cc-cli/src/commands/collectCoins.ts:
```diff
@@ -20,7 +18,19 @@ export function makeCollectCoinsCmd() {
+function collectCoinsActionSync(options: OptionValues) {
+  collectCoinsAction(options)
+    .then(() => {
+      console.log("Success");
```
```suggestion
      console.log(chalk.green("Success!"));
```

as in the removed `handleSuccess`.

atodorov commented on scripts/cc-cli/src/commands/collectCoins.ts:
```diff
@@ -20,7 +18,19 @@ export function makeCollectCoinsCmd() {
+    })
+    .catch((reason) => {
+      console.log(reason);
+      process.exit(1);
```
```
fatalErr(
    `ERROR: The call to request_collect_coins was unsuccessful: ${
      reason as string
    }`,
```

as in the removed `handleError` implementation

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,119 @@
+      caller = randomTestAccount(false);
+
+      ccApi = await creditcoinApi((global as any).CREDITCOIN_API_URL);
+      collector = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "lender");
```
Can rename collector -> sudo

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,119 @@
+        parseAmountInternal("1000000"),
+        arg("CREDITCOIN_API_URL"),
+      );
+      await signSendAndWatch(fundTx, api, initKeyringPair("//Alice"));
```
`initKeyringPair("//Alice")` -> that's `sudo` from above, isn't it ?

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,119 @@
+      ).toBe(true);
+
+      const collectResult = execa.commandSync(
+        `npx creditcoin-cli collect-coins -u ${
```
Is the `npx` command installed as part of dependencies? I remember I didn't have it when I was working on something else.

zacharyfrederick replied:
`npx` is installed by default with `npm` version > 5.2.0.

atodorov commented on scripts/cc-cli/src/test/integration-tests/registerAddress.test.ts:
```diff
@@ -0,0 +1,69 @@
+        parseAmountInternal("1000000"),
+        arg("CREDITCOIN_API_URL"),
+      );
+      await signSendAndWatch(fundTx, api, initKeyringPair("//Alice"));
```
Alice -> arg("CREDITCOIN_CREATE_SIGNER")(keyring, "lender");

atodorov reviewed (changes requested):
Added some more comments and minor issues.

Commit:
moved token deployment to beforeAll method of the collect coins test for the cli

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
formatted

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,138 @@
+    caller = randomTestAccount(false);
+
+    ccApi = await creditcoinApi((global as any).CREDITCOIN_API_URL);
+    sudo = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "lender");
```
```suggestion
    sudo = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "sudo");
```

atodorov commented on scripts/cc-cli/src/test/integration-tests/registerAddress.test.ts:
```diff
@@ -0,0 +1,76 @@
+      arg("CREDITCOIN_CREATE_WALLET"),
+    );
+
+    sudo = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "lender");
```
```suggestion
    sudo = arg("CREDITCOIN_CREATE_SIGNER")(keyring, "sudo");
```

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,138 @@
+    expect(collectResult.failed).toBe(false);
+    expect(collectResult.exitCode).toBe(0);
+    expect(collectResult.stderr).toBe("");
+    expect(collectOutput[collectOutput.length - 1]).toBe("Success!");
```
Can you assert on the available balance before and after collect coins ?

zacharyfrederick replied:
This one is proving trickier. 

Checking the balance before and after using Pablo's `getBalance` call shows a `0` difference in transferable tokens. I see the mint event happening and we have tested this previously so I must be doing something wrong. Let me keep working on this one

zacharyfrederick replied:
`api.query.system.account` also shows values that don't make sense...

atodorov replied:
> `api.query.system.account` also shows values that don't make sense...

Anything in particular? 

I'm not able to find documentation for this query, only found this https://polkadot.js.org/docs/api/FAQ#my-chain-does-not-support-systemaccount-queries, so I'm assuming the balances don't look right.


Maybe you can use `getBalance()`, see https://polkadot.js.org/docs/substrate/rpc/#getbalanceaddress-h160-number-blocknumber-u256. 

If you can push some WIP code of that you are trying to do I will take a look.

atodorov replied:
I just realized we can also use `getBalance()` from cc-cli, see https://github.com/gluwa/creditcoin/blob/dev/scripts/cc-cli/src/utils/balance.ts. 

Does that show what you would expect ?

zacharyfrederick replied:
Using the `api.query.system.account` method showed a difference of 0. 

The `getBalance` command also showed zero. I pushed the changes that show how I am making those calls and asserting on them

zacharyfrederick replied:
In addition to the method tried in the latest push I also tried something similar to this `expect(ending.total.sub(starting.total).eq(new BN(1))).toBe(true);` without success

atodorov commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,138 @@
+    expect(registerResult.stderr).toBe("");
+    expect(
+      stdout[stdout.length - 1].includes("Address Registered Successfully"),
+    ).toBe(true);
```
Since this is `collectCoins` you don't really need to assert on the output of register-address here (that's done explicitly in `registerAddress.test.ts` anyway. 

You can also go ahead and use `tryRegisterAddress` from creditcoin-js

atodorov reviewed (changes requested):
Minor nitpicks, mainly for house-keeping, otherwise looks good.

Commit:
replaced cli call with tryRegisterAddress

Commit:
removed console.log statement

Commit:
added wip code to assert on balance

atodorov committed:
Patch failing assertion inside collect-coins test

- be more verbose when calling .burn() on the token (not related to the
  actual failure)
- increase the amount we burn b/c it was too low and the difference was
  obscured by collected transaction fees
- use different numbers b/c "1" followed by multiple zeroes is hard to read

- sleep 3 seconds before fetching ending balance otherwise it doesn't work

nathanwhit commented on scripts/cc-cli/src/commands/collectCoins.ts:
```diff
@@ -31,9 +45,12 @@ async function collectCoinsAction(options: OptionValues) {
+    signer,
+    options.burnTxHash,
+  );
+  await event.waitForVerification(800_000);
```
```suggestion
  await event.waitForVerification(900_000);
```

We currently guarantee that a given collect coins will be dealt with in 60 blocks (after that the task expires and you'd have to re-submit the transaction). So in the worst case `60 * 15s = 900_000`

zacharyfrederick replied:
This change and the others mentioned are addressed in #1326. Please keep the conversation unresolved for easier reference :)

nathanwhit commented on scripts/cc-cli/src/globalSetup.ts:
```diff
@@ -0,0 +1,134 @@
+  await api.disconnect();
+}
+
+export function retry(retries: any, executor: any): any {
+  console.log(`${retries as string} retries left!`);
+
+  if (typeof retries !== "number") {
+    throw new TypeError("retries is not a number");
+  }
+
+  return new Promise(executor).catch((error) =>
+    retries > 0 ? retry(retries - 1, executor) : Promise.reject(error),
+  );
+}
```
```suggestion
export async function retry<T>(
  retries: number,
  func: () => PromiseLike<T>
): Promise<T> {
  console.log(`${retries} retries left!`);

  try {
    return await func();
  } catch (error) {
    if (retries > 0) {
      return await retry(retries - 1, func);
    }
    throw error;
  }
}
```

IMO it's only necessary to drop down to using `new Promise` and explicit resolve/reject callbacks when you need more control over the promise, such as when you're converting a callback-based API to promises (which is what we do in creditcoin-js, since polkadot-js uses observables which are effectively callback-based). If the caller needs that, then they can do it themselves.

99% of the time though you don't need that, and you just want to use normal async/await notation.

nathanwhit commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,142 @@
+      process.env.CREDITCOIN_CTC_BURN_TX_HASH,
+    );
+
+    await retry(5, async (resolve: any, reject: any) => {
+      await setAuthorities().then(resolve).catch(reject);
+    }).catch(() => {
```
```suggestion
    await retry(5, setAuthorities).catch(() => {
```

If you choose to accept my suggestion for the retry function, on the caller side it would just look like this

nathanwhit commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,142 @@
+
+    await retry(5, async (resolve: any, reject: any) => {
+      await setAuthorities().then(resolve).catch(reject);
+    }).catch(() => {
+      console.log("Could not setup testing authorities");
+      process.exit(1);
+    });
```
This could be a normal `try`-`catch` instead of using `.catch`. In general, I only use `.catch` and `.then` if it's in a sync context (i.e. there's no async keyword and you aren't allowed to use `await`). Since this is already running in an `async` function, you might as well use try-catch.

nathanwhit commented on scripts/cc-cli/src/test/integration-tests/collectCoins.test.ts:
```diff
@@ -0,0 +1,142 @@
+    expect(collectOutput[collectOutput.length - 1]).toBe("Success!");
+
+    // note: wait before fetching the balance again. Otherwise we get bogus values
+    await new Promise((f) => setTimeout(f, 3000));
```
I thought about it some more and this still feels weird to me. I don't have any answers, but really as soon as the collect coins has been verified the balance should reflect the minted coins. So unless polkadot JS is quietly using finalized state or something, or there's a bug in polkadot JS, it doesn't really make sense to me that you would be getting stale values.

Anything involving polkadot JS internals is a bit of a (massive) rabbit hole though, so I'm okay with leaving this a mystery for now. But FTR it still doesn't sit right with me that we have to wait like this at all

atodorov replied:
FTR this is already converted to waiting for a number of blocks (2), which is better than an arbitrary amount in seconds: https://github.com/gluwa/creditcoin/pull/1325/commits/b62fa811594cbf4a2bca6309cf201604bd055053

nathanwhit reviewed (commented):
A few typescript best practices nitpicks but otherwise looks good to me
